src/schema.rs

```diff
@@ -52,9 +51,14 @@ impl Schema {
 
     /// Returns an iterator over the types in this schema.
     pub(crate) fn get_types(&self) -> SchemaTypeIterator {
-        SchemaTypeIterator {
-            types: self.types.get_types().iter(),
-        }
+        SchemaTypeIterator::new(
+            self.types
+                .get_types()
+                .iter()
+                .enumerate()
+                .map(|(i, t)| TypeRef::new(i, Rc::clone(&self.types)))
+                .collect(),
+        )
```

Comment:
Iterators are usually lazy so the caller only pays for what they use. Allocating a `Vec` and building all of the `TypeRef`s when `get_types()` is called would be pretty expensive if the caller only wanted the first type in the schema.

Your original implementation was very close to what I had in mind; `SchemaTypeIterator` should have a lifetime linked to `&self` and, instead of passing out `&TypeDefinition`s, would construct a single `TypeRef` each time `next()` was called.

Reply:
The issue here was when trying to convert `TypeDefinition` into `TypeRef` instead of collecting into vector, I canuse `.iter()` but that creates a temporary value created in this function hence can not return it. So I had to change it to rather be `Vec<TypeRef>`.  Did you have something particular in mind for this?

Reply:
Changes `SchemaTypeIterator` to store `Rc<TypeStore> `and create `TypeRef` when `.next()` is called.